### PR TITLE
Extract dup and idup to own module

### PR DESCRIPTION
**Pull request by dkorpel**
Make object.d smaller by extracting the implementation and leaving a public import. See:

https://github.com/dlang/druntime/pull/3860#issuecomment-1170139760

I'm not sure core/internal/array is the right folder, since it's not the compiler that's 'lowering' the `.dup` property to a function anymore, it's just a normal function at this point. Suggestions for a better location are welcome.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/core/internal/array/duplicate.d [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+    assert(s == "abc");
+}
+
+private U[] _dup(T, U)(scope T[] a) pure nothrow @trusted if (__traits(isPOD, T))
+{
+    if (__ctfe)
+        return _dupCtfe!(T, U)(a);
+
+    import core.stdc.string : memcpy;
+    import rt.lifetime : _d_newarrayU;
```
Comment (dkorpel) [inline src/core/internal/array/duplicate.d:140]: This is the only line I added, the rest is copy-paste

**Review by ljmf00 — APPROVED** [review]

**Comment by dkorpel** [comment]
Looks like this is not as simple as I thought, I get hard to decipher errors about attributes, `cast(void*)0` at CTFE, circular module constructor dependencies etc. I'll start by extracting the tests and work incrementally from there.

**Comment by dkorpel** [comment]
Some things to take into account:
- DMD has a special CTFE builtin for `dup` in `object.d`.
- object.d has `extern(C)` declarations for `_aaGetX` and `_d_newarrayU` with fake `pure` `nothrow` attributes

File: src/object.d [hunk 2]
```diff
@@ -4085,209 +4070,6 @@ private void _doPostblit(T)(T[] arr)
...
-        this(this) { gp = p; }
-    }
-
-    int p;
-    scope S[1] arr = [S(&p)];
-    auto a = arr.dup; // dup does escape
-}
-
 // https://issues.dlang.org/show_bug.cgi?id=21983
 // dup/idup destroys partially constructed arrays on failure
```
Comment (ibuclaw) [inline src/object.d:4074]: This unittest too.

**Comment by adamdruppe** [comment]
> Make object.d smaller

What's the user-visible goal this is supposed to achieve? Have you confirmed it actually makes it?

**Comment by dkorpel** [comment]
> What's the user-visible goal this is supposed to achieve?

> See:
> [#3860 (comment)](https://github.com/dlang/druntime/pull/3860#issuecomment-1170139760)

**Comment by adamdruppe** [comment]
If the goal is to make the other name user visible, we'd probably have to use something other than `core.internal`...

**Comment by dkorpel** [comment]
> If the goal is to make the other name user visible, we'd probably have to use something other than `core.internal`...

I don't understand the folder structure of druntime (why are some compiler hooks in `core/internal` and others in `rt/`?), so I'm open to suggestions.

**Comment by adamdruppe** [comment]
`rt` is not accessible once druntime is compiled; it isn't included in the `druntime/import` that actually ships. So it can only be used through `extern(C)` hacks/compiler hooks, or pure internal druntime use. It can never be `public import`ed since user code can't see it at all.

`core.internal` can be public imported from other `core` modules, but are not supposed to be imported or otherwise referenced by end user modules, since they are not part of the stable, published api. If a `core.internal` thing is public imported, that's a pure implementation detail - users are still supposed to reference it through the `object` namespace (or whatever core.x thing public imported it).

**Comment by RazvanN7** [comment]
I think that if we were to follow the convention, the file should be named `duplication.d`.

**Comment by ljmf00** [comment]
> Some things to take into account:
> 
> * DMD has a special CTFE builtin for `dup` in `object.d`.

I get really annoyed by these special cases. I know that `object` is special on what I'm going to say but, we have builtins hardcoded all over the place. `core.math`, `std.math` and `object` has to have a bunch of "dummy" signatures because of these special cases on the compiler. Even the compiler needs to have special logic/branching for `Id.std`, `Id.core` and other special modules/packages because of these hardcoded builtins. Moving forward with a unified way to get/use builtins from the compiler is the way GCC and LLVM (`__builtin_` prefix) already goes. Since we have modules, we can do it in a cleaner way, like LDC does with attributes and compiler-specific intrinsics.

**Comment by maxhaton** [comment]
> > Some things to take into account:
> > 
> > * DMD has a special CTFE builtin for `dup` in `object.d`.
> 
> I get really annoyed by these special cases. I know that `object` is special on what I'm going to say but, we have builtins hardcoded all over the place. `core.math`, `std.math` and `object` has to have a bunch of "dummy" signatures because of these special cases on the compiler. Even the compiler needs to have special logic/branching for `Id.std`, `Id.core` and other special modules/packages because of these hardcoded builtins. Moving forward with a unified way to get/use builtins from the compiler is the way GCC and LLVM (`__builtin_` prefix) already goes. Since we have modules, we can do it in a cleaner way, like LDC does with attributes and compiler-specific intrinsics.

I agree. it's really stupid

**Review by RazvanN7 — APPROVED** [review]
LGTM

**Pull request merged**